### PR TITLE
⚠️ Generate v1alpha4 types 

### LINKED ISSUE #1969: Create v1alpha4 types, conversion webhooks from v1alpha3 to v1alpha4 and remove v1alpha2
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

CAPA 0.7.0 will be v1alpha4, and therefore api/v1alpha4 types need to be created, as well as conversion webhooks from v1alpha3 to v1alpha4, and the v1alpha2 types deleted.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: v0.7.0
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sedefsavas**
/kind api-change
**What this PR does / why we need it**:
- Delete v1alpha2 types
- Create v1alpha4 types
- Add conversion hubs for bootstrap/ and controlplane/ and exp/ APIs
- Update controller-runtime to v0.9.x 
- Update golang to v1.16.2
- Update leader-election RBAC rules following: https://github.com/kubernetes-sigs/cluster-api/pull/3756/files
- Update admission webhook in line with: https://github.com/kubernetes-sigs/cluster-api/pull/3753/files
- Uses the CAPI nightly builds for E2E tests with v1alpha4 types

There are many changes we want to add in v1alpha4 but the purpose of this PR is just to open the main branch to v1alpha4 changes, we can start merging all the changes for v1alpha4 after that.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1969, https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2076

**Release note**:

```release-note
Generate v1alpha4 API version types and remove v1alpha2 types
```

### CODE REVIEW

**Comment by sedefsavas** [comment]
TODO:

**Comment by sedefsavas** [comment]
This PR is ready for review now. Since new PRs may require significant change in this api creation, hoping to merge this to master soon. For example, even bumping controller-runtime from v0.7.0-alpha.5 to v0.7.0-alpha.8 required bunch of changes.
cc @randomvariable @richardcase @MarcusNoble 

e2e tests are failing because CAPI does not have a release with v1alpha4 types yet (there is a plan to release one this week, will update the e2e-conf.yaml accordingly after that).

`make verify` is locally passing, but it is failing on prow due to `kubekins-e2e:v20201111-a263fd7-1.19` we are using in our prow tests, somewhere it has an earlier version of `github.com/googleapis/gnostic` and that version has this file:` OpenAPIv2.go`, whereas newer versions of gnostic has openapiv1, so getting this error:

`        github.com/googleapis/gnostic/OpenAPIv2: module github.com/googleapis/gnostic@latest found (v0.5.3), but does not contain package github.com/googleapis/gnostic/OpenAPIv2
`

Couldn't find how to solve this yet.

**Comment by richardcase** [comment]
Great, will take a look in the morning.

File: controlplane/eks/api/v1alpha4/types.go
Comment (MarcusNoble) [inline controlplane/eks/api/v1alpha4/types.go:37]:
```suggestion
	// ControllerManager indicates if the controller manager (kube-controller-manager) log should be enabled
```

File: exp/api/v1alpha4/types.go
Comment (MarcusNoble) [inline exp/api/v1alpha4/types.go:53]:
```suggestion
	EBS EBS `json:"ebs,omitempty"`
```

File: exp/api/v1alpha4/types.go
Comment (MarcusNoble) [inline exp/api/v1alpha4/types.go:51]: Should this say `DeviceName` instead of `VirtualName` ?

File: exp/api/v1alpha4/types.go
Comment (MarcusNoble) [inline exp/api/v1alpha4/types.go:99]: Should this support adding non-root / secondary volumes too?

File: exp/api/v1alpha4/types.go
Comment (MarcusNoble) [inline exp/api/v1alpha4/types.go:92]: Same comments about image lookup and AMI. Interesting that this one has the option for providing both. Is `AMI` used over the image lookup if provided?

File: hack/ensure-kubectl.sh
Comment (MarcusNoble) [inline hack/ensure-kubectl.sh:22]: How come we're dropping the min version?

**Review by MarcusNoble — CHANGES_REQUESTED** [review]

File: hack/ensure-kubectl.sh
Comment (sedefsavas) [reply hack/ensure-kubectl.sh]: This belongs to the `revert later` commit but you are right, we only need to revert go version, not KUBECTL min version. Changed this.

As a side note to I will revert golang min version once I update prow images from 1.18 to 1.19 (because 1.19 is using go 1.15): https://github.com/kubernetes/test-infra/blob/9d2209384daf45ceef4016a4f1f04d06d37f48bf/config/jobs/kubernetes-sigs/cluster-api-provider-aws/cluster-api-provider-aws-periodics.yaml#L20

File: api/v1alpha4/awscluster_types.go
Comment (sedefsavas) [reply api/v1alpha4/awscluster_types.go:33]: There are many changes we want to add in v1alpha4 but the purpose of this PR is just to open the main branch to v1alpha4 changes, we can start merging all the changes for v1alpha4 after that.

File: api/v1alpha4/awscluster_types.go
Comment (sedefsavas) [reply api/v1alpha4/awscluster_types.go:80]: Same as my previous comment.

File: api/v1alpha4/awscluster_types.go
Comment (sedefsavas) [reply api/v1alpha4/awscluster_types.go:111]: Same comment as above.
Did not do any changes on the v1alpha4 types.
Let's track the changes we want to have in v1alpha4 as separate issues so that it will be easier to review/revert.

File: api/v1alpha4/awsmachine_types.go
Comment (sedefsavas) [reply api/v1alpha4/awsmachine_types.go:118]: Same comment as above.
Did not do any changes on the v1alpha4 types.
Let's track the changes we want to have in v1alpha4 as separate issues so that it will be easier to review/revert.

File: bootstrap/eks/controllers/eksconfig_controller.go
Comment (sedefsavas) [reply bootstrap/eks/controllers/eksconfig_controller.go:293]: I am introducing any changes in v1alpha4 types.
Let's track these in separate issues.

Comment (sedefsavas) [reply bootstrap/eks/controllers/eksconfig_controller.go:293]: Ah good question, this is due to the new controller runtime changes. I am following the changes in CAPI v1alpha4 related to this: https://github.com/kubernetes-sigs/cluster-api/pull/3735

File: controllers/awscluster_controller.go
Comment (sedefsavas) [reply controllers/awscluster_controller.go:66]: Yes they are coming from the controller runtime:
https://github.com/kubernetes-sigs/controller-runtime/blob/5f112fb2c95a609e182e720d0c7ec1ce693717e0/pkg/internal/controller/controller.go#L92
```
// Reconcile implements reconcile.Reconciler
func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
	log := c.Log.WithValues("name", req.Name, "namespace", req.Namespace)
	ctx = logf.IntoContext(ctx, log)
	return c.Do.Reconcile(ctx, req)
}
```

File: api/v1alpha4/conditions_consts.go
Comment (sedefsavas) [reply api/v1alpha4/conditions_consts.go:29]: Same as above.

File: controlplane/eks/api/v1alpha4/types.go
Comment (sedefsavas) [reply controlplane/eks/api/v1alpha4/types.go:37]: Same as above

File: exp/api/v1alpha4/types.go
Comment (sedefsavas) [reply exp/api/v1alpha4/types.go:53]: Same as above.

File: exp/api/v1alpha4/types.go
Comment (sedefsavas) [reply exp/api/v1alpha4/types.go:51]: It may me hard to create separate issue for all these little things, so I created an issue and pointing to this PR, so once master is v1alpha4, we can target these.

https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2157

File: exp/api/v1alpha4/types.go
Comment (sedefsavas) [reply exp/api/v1alpha4/types.go:99]: Tracking here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2157

File: exp/api/v1alpha4/types.go
Comment (sedefsavas) [reply exp/api/v1alpha4/types.go:92]: Tracking here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2157

File: api/v1alpha4/awscluster_types.go
Comment (MarcusNoble) [reply api/v1alpha4/awscluster_types.go:33]: Wasn't aware of that 😅 Most of my comments can be ignored then 😆

Comment (sedefsavas) [reply api/v1alpha4/awscluster_types.go:33]: My bad, updated the description.

Comment (sedefsavas) [reply api/v1alpha4/awscluster_types.go:33]: I am not resolving these and tracking because we want to do these changes after we merge.

File: bootstrap/eks/controllers/eksconfig_controller.go
Comment (MarcusNoble) [reply bootstrap/eks/controllers/eksconfig_controller.go:293]: Might be worth raising this either on Slack or at the office hours on Monday to get a general direction.

Comment (sedefsavas) [reply bootstrap/eks/controllers/eksconfig_controller.go:293]: Okay, added to the agenda for Monday.

**Comment by dongsupark** [comment]
Hi! :wave: 
Thanks for the great work.

However, can you please add `cluster.x-k8s.io/v1alpha4: v1alpha4` to [config/crd/kustomization.yaml](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/68bdb89eb9ecb88a5ed0eb06b37cbe5077514cfe/config/crd/kustomization.yaml#L2)?

Without that change, the `capi-controller-manager` pod would stuck with the following error messages:

```
E1216 14:49:42.349223       1 machinedeployment_controller.go:144] controller/machinedeployment
"msg"="Failed to reconcile MachineDeployment"
"error"="cannot find any versions matching contract \"cluster.x-k8s.io/v1alpha4\" for CRD awsmachinetemplates.infrastructure.cluster.x-k8s.io"
"name"="capi-quickstart-md-0" "namespace"="default"
"reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachineDeployment"
```

**Comment by randomvariable** [comment]
I've changed the base of this PR to release-0.7 branch

**Comment by sedefsavas** [comment]
/hold

until daily build PR is merged: https://github.com/kubernetes-sigs/cluster-api/pull/4101

**Comment by invidian** [comment]
>until daily build PR is merged: kubernetes-sigs/cluster-api#4101

This is now merged.

File: api/v1alpha3/zz_generated.conversion.go
Comment (sbueringer) [inline api/v1alpha3/zz_generated.conversion.go:445]: In case you're not already aware anyway. CAPZ solved it like this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/55c757c661f689fff8fb818e7171472ad37122a1/api/v1alpha3/azurecluster_conversion.go#L221

**Comment by sedefsavas** [comment]
/hold
until https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2253 is merged given that multitenancy PR has major API changes.

File: api/v1alpha3/zz_generated.conversion.go
Comment (sedefsavas) [reply api/v1alpha3/zz_generated.conversion.go:445]: If this is about compileErrorOnMissingConversion() error during generation conversion, it is solved by changing pointer slices to slices.
https://github.com/kubernetes-sigs/cluster-api/issues/4096#issuecomment-765597734

Or is this another issue?

Comment (sbueringer) [reply api/v1alpha3/zz_generated.conversion.go:445]: I think it's about providing conversion functions for types which are not in the local go module.
I.e. in CAPZ they provided Convert functions for APIEndpoint (APIEndpoint is from the cluster-api) repo in the same package and conversion-gen then just used them in the generated code.

Afaik APIEndpoint is not used in slices (https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/api/v1alpha3/awscluster_types.go#L44) so it should be another issue.

Just stumpled upon the issue when implementing CAPO v1alpha4 and saw the TODO here and the (probably) correct implementation in CAPZ and though I leave you a hint here.

**Comment by sedefsavas** [comment]
TODO left for this PR:

1. Daily-built images will be added to e2e tests.
2. EKS controlplane webhook will be combined into controller manager deployment, similar to core CAPA controller.

The rest is ready for the review.
cc @randomvariable @richardcase

**Comment by sedefsavas** [comment]
/hold cancel

This is ready for review.
/cc @richardcase

**Comment by richardcase** [comment]
@sedefsavas @randomvariable  - i'll have a look now

File: bootstrap/eks/api/v1alpha3/doc.go
Comment (richardcase) [inline bootstrap/eks/api/v1alpha3/doc.go:2]: Guessing this move was intended?

File: bootstrap/eks/api/v1alpha4/eksconfig_types.go
Comment (richardcase) [inline bootstrap/eks/api/v1alpha4/eksconfig_types.go:62]: Does this need to be removed?

File: bootstrap/eks/api/v1alpha4/eksconfigtemplate_types.go
Comment (richardcase) [inline bootstrap/eks/api/v1alpha4/eksconfigtemplate_types.go:35]: And maybe this?

**Comment by richardcase** [comment]
This looks good to me.....some very minor comments. 

The EKS e2e test are failing. I'd be happy to fix these as part #2355 if you want?

File: api/v1alpha4/awsmachine_types.go
Comment (randomvariable) [reply api/v1alpha4/awsmachine_types.go:17]: I suspect it's an artifact of how git interpreted the deletion of v1alpha2 and v1alpha3's continued existence.

Comment (randomvariable) [reply api/v1alpha4/awsmachine_types.go:17]: Maybe easier to do as separate commits, so the git history is easier.

**Comment by randomvariable** [comment]
I'll fix my messup in test-infra

File: bootstrap/eks/api/v1alpha4/eksconfig_types.go
Comment (sedefsavas) [reply bootstrap/eks/api/v1alpha4/eksconfig_types.go:62]: We need this so that we store the api in this version.
Removed this from v1alpha3 types.

File: bootstrap/eks/api/v1alpha4/eksconfigtemplate_types.go
Comment (sedefsavas) [reply bootstrap/eks/api/v1alpha4/eksconfigtemplate_types.go:35]: We need this so that we store the api in this version.
Removed this from v1alpha3 types.

**Comment by sedefsavas** [comment]
go version in the docs build looks 1.14.4, not sure where that is set.
cc @randomvariable

File: bootstrap/eks/api/v1alpha3/doc.go
Comment (sedefsavas) [reply bootstrap/eks/api/v1alpha3/doc.go]: Fixed.

**Comment by sedefsavas** [comment]
/test pull-cluster-api-provider-aws-e2e-eks

One of the problems, eks test were not passing was e2e clean-up: in e2e tests, we need to use cluster-api `DeleteCluster` instead of `DeleteClusterAndWait` because the global resources remain and it fails.

**Review by richardcase — APPROVED** [review]
This last commit looks good to me so:

/lgtm

**Pull request merged**
